spring-cloud-netflix-eureka-server/src/main/java/org/springframework/cloud/netflix/eureka/server/EurekaServerAutoConfiguration.java

```diff
@@ -169,9 +172,69 @@ public PeerAwareInstanceRegistry peerAwareInstanceRegistry(
 	@ConditionalOnMissingBean
 	public PeerEurekaNodes peerEurekaNodes(PeerAwareInstanceRegistry registry,
 			ServerCodecs serverCodecs) {
-		return new PeerEurekaNodes(registry, this.eurekaServerConfig,
+		return new RefreshablePeerEurekaNodes(registry, this.eurekaServerConfig,
 				this.eurekaClientConfig, serverCodecs, this.applicationInfoManager);
 	}
+	
+	/**
+	 * {@link PeerEurekaNodes} which updates peers when /refresh is invoked.
+	 * Peers are updated only if
+	 * <code>eureka.client.use-dns-for-fetching-service-urls</code> is
+	 * <code>false</code> and one of following properties have changed.
+	 * </p>
+	 * <ul>
+	 * <li><code>eureka.client.availability-zones</code></li>
+	 * <li><code>eureka.client.region</code></li>
+	 * <li><code>eureka.client.service-url.&lt;zone&gt;</code></li>
+	 * </ul>
+	 */
+	static class RefreshablePeerEurekaNodes extends PeerEurekaNodes
+			implements ApplicationListener<EnvironmentChangeEvent> {
+
+		public RefreshablePeerEurekaNodes(
+				final PeerAwareInstanceRegistry registry,
+				final EurekaServerConfig serverConfig,
+				final EurekaClientConfig clientConfig, 
+				final ServerCodecs serverCodecs,
+				final ApplicationInfoManager applicationInfoManager) {
+			super(registry, serverConfig, clientConfig, serverCodecs, applicationInfoManager);
+		}
+
+		@Override
+		public void onApplicationEvent(final EnvironmentChangeEvent event) {
+			if (shouldUpdate(event.getKeys())) {
+				updatePeerEurekaNodes(resolvePeerUrls());
+			}
+		}
+		
+		/*
+		 * Check whether specific properties have changed.
+		 */
+		protected boolean shouldUpdate(final Set<String> changedKeys) {
+			assert changedKeys != null;
+			
+			if (clientConfig.shouldUseDnsForFetchingServiceUrls()) {
+				return false;
+			}
+			
+			if (changedKeys.contains("eureka.client.region")) {
+				return true;
+			}
+			
+			for (final String key : changedKeys) {
+				assert key != null;
```

Comment:
Why throw an exception? Use an if.

Reply:
I used an `assert` here only to specify the assumption that property keys cannot be null, hence I'm not doing a null check before calling `key.startsWith()`. Since assertions are disabled in production, hope this should fine?